notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
+def get_notebooks():
+    user_id = request.headers.get('User')
 
+    if is_falsy_str(user_id):
+        return forbidden()
 
-def get_all_workers():
-    workers = k8s.list_namespaced_pod(
+    # Not well-documented: empty selector returns nothing
+    # https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/CoreV1Api.md
+    # https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
+    pods = k8s.list_namespaced_pod(
         namespace='default',
-        watch=False,
-        label_selector='app=notebook-worker',
-        _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-    workers_and_svcs = []
-    for w in workers.items:
-        uuid = w.metadata.labels['uuid']
-        svcs = k8s.list_namespaced_service(
-            namespace='default',
-            watch=False,
-            label_selector='uuid=' + uuid,
-            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS).items
-        assert len(svcs) <= 1
-        if len(svcs) == 1:
-            workers_and_svcs.append((w, svcs[0]))
-        else:
-            log.info(f'assuming pod {w.metadata.name} is getting deleted '
-                     f'because it has no service')
-    return workers_and_svcs
-
-
-@app.route('/workers')
-def workers():
-    if not session.get('admin'):
-        return redirect(external_url_for('admin-login'))
-    workers_and_svcs = get_all_workers()
-    return render_template('workers.html',
-                           workers=workers_and_svcs,
-                           workers_url=external_url_for('workers'),
-                           leader_instance=INSTANCE_ID)
-
-
-@app.route('/workers/<pod_name>/<svc_name>/delete')
-def workers_delete(pod_name, svc_name):
-    if not session.get('admin'):
-        return redirect(external_url_for('admin-login'))
-    delete_worker_pod(pod_name, svc_name)
-    return redirect(external_url_for('workers'))
-
-
-@app.route('/workers/delete-all-workers', methods=['POST'])
-def delete_all_workers():
-    if not session.get('admin'):
-        return redirect(external_url_for('admin-login'))
-    workers_and_svcs = get_all_workers()
-    for pod_name, svc_name in workers_and_svcs:
-        delete_worker_pod(pod_name, svc_name)
-    return redirect(external_url_for('workers'))
-
-
-def delete_worker_pod(pod_name, svc_name):
-    try:
-        k8s.delete_namespaced_pod(
-            pod_name,
-            'default',
-            kube.client.V1DeleteOptions(),
-            _request_timeout=KUBERNETES_TIMEOUT_IN_SECONDS)
-    except kube.client.rest.ApiException as e:
-        log.info(f'pod {pod_name} or associated service already deleted {e}')
+        label_selector=f"user_id={UNSAFE_user_id_transform(user_id)}",
+        timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
+
+    return marshall_json(pods, pod_paths), 200
+
+
+@app.route('/jupyter/<svc_name>', methods=['DELETE'])
+def delete_notebook(svc_name: str):
+    # TODO: Is it possible to have a falsy token value and get here?
```

Comment:
No, right? Otherwise the authentication service has made a mistake.

Reply:
Notebook has no control over the authorization system (please lets stop calling it authentication).